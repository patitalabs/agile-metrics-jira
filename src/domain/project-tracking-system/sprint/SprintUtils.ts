import { SprintHealthEstimate } from './SprintHealthEstimate';
import { Utils } from '../../../utils/Utils';

export class SprintUtils {
  static cycleTime(movedToDev: Date, resolutionDate: Date): number {
    let cycleTime = 0;
    if (movedToDev) {
      let endLeadTime = new Date();
      if (resolutionDate) {
        endLeadTime = resolutionDate;
      }
      cycleTime = Utils.daysBetween(movedToDev, endLeadTime);
    }
    return cycleTime;
  }

  static leadTime({ created, resolutionDate }): number {
    let leadTime = 0.0;
    if (created) {
      let endLeadTime = new Date();
      if (resolutionDate) {
        endLeadTime = resolutionDate;
      }
      leadTime = Utils.daysBetween(created, endLeadTime);
      if (leadTime === 0.0) {
        leadTime = 1;
      }
    }
    return leadTime;
  }

  static estimateHealth({
    estimate,
    actualTime,
    maxTime,
    estimationValues,
  }): number {
    if (!estimationValues) {
      estimationValues = [1, 2, 3, 5, 8];
    }
    return new SprintHealthEstimate({
      maxTime,
      estimationValues,
      estimate,
      actualTime,
    }).calculateHealthFactor();
  }

  static movedToDev(movedForwardDates: Date[]): Date {
    let movedToDev: Date = null;
    if (movedForwardDates.length > 0) {
      movedToDev = movedForwardDates.reduce((dateOne, dateTwo) =>
        dateOne < dateTwo ? dateOne : dateTwo
      );
    }
    return movedToDev;
  }
}
