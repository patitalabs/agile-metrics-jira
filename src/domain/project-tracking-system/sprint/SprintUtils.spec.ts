import { SprintUtils } from './SprintUtils';

describe('SprintUtils', () => {
  it('cycleTime', () => {
    const cycleTime = SprintUtils.cycleTime(
      new Date('2018-12-04'),
      new Date('2018-12-14')
    );
    expect(cycleTime).toBe(10);
  });

  it('leadTime', () => {
    const leadTime = SprintUtils.leadTime({
      created: new Date('2018-12-02'),
      resolutionDate: new Date('2018-12-12'),
    });
    expect(leadTime).toBe(10);
  });

  it('movedToDev', () => {
    const movedToDev = SprintUtils.movedToDev([
      new Date('2018-12-03'),
      new Date('2018-12-13'),
    ]);
    expect(movedToDev).toMatchSnapshot();
  });

  it('estimateHealth underEstimate', () => {
    const estimateHealth = SprintUtils.estimateHealth({
      estimate: 2,
      actualTime: 6,
      maxTime: 8,
      estimationValues: [1, 2, 3, 5, 8],
    });
    expect(estimateHealth).toBe(3.5);
  });

  it('estimateHealth exact estimate', () => {
    const estimateHealth = SprintUtils.estimateHealth({
      estimate: 5,
      actualTime: 5,
      maxTime: 13,
      estimationValues: [1, 2, 3, 5, 8, 13],
    });
    expect(estimateHealth).toBe(0);
  });

  it('estimateHealth oveEstimate', () => {
    const estimateHealth = SprintUtils.estimateHealth({
      estimate: 5,
      actualTime: 1,
      maxTime: 8,
      estimationValues: [1, 2, 3, 5, 8],
    });
    expect(estimateHealth).toBe(-3);
  });

  it('estimateHealth underEstimate no estimate', () => {
    const estimateHealth = SprintUtils.estimateHealth({
      estimate: 1,
      actualTime: 6,
      maxTime: 3,
      estimationValues: [1],
    });
    expect(estimateHealth).toBe(3);
  });

  it('estimateHealth exact estimate no estimate', () => {
    const estimateHealth = SprintUtils.estimateHealth({
      estimate: 1,
      actualTime: 1,
      maxTime: 14,
      estimationValues: [1],
    });
    expect(estimateHealth).toBe(0);
  });

  it('estimateHealth oveEstimate no estimate', () => {
    const estimateHealth = SprintUtils.estimateHealth({
      estimate: 1,
      actualTime: 0,
      maxTime: 14,
      estimationValues: [1],
    });
    expect(estimateHealth).toBe(-1);
  });
});
