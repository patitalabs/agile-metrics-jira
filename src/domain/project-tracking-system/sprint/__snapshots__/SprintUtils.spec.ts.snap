// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SprintUtils movedToDev 1`] = `2018-12-03T00:00:00.000Z`;
